src/node.cc

```diff
@@ -1015,65 +1029,71 @@ InitializationResult InitializeOncePerProcess(int argc, char** argv) {
+      if (credentials::SafeGetenv("NODE_EXTRA_CA_CERTS", &extra_ca_certs))
+        crypto::UseExtraCaCerts(extra_ca_certs);
     }
-    OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CONFIG, settings);
-    OPENSSL_INIT_free(settings);
-
-    if (ERR_peek_error() != 0) {
-      result.exit_code = ERR_GET_REASON(ERR_peek_error());
-      result.early_return = true;
-      fprintf(stderr, "OpenSSL configuration error:\n");
-      ERR_print_errors_fp(stderr);
-      return result;
+    // In the case of FIPS builds we should make sure
+    // the random source is properly initialized first.
+#if OPENSSL_VERSION_MAJOR >= 3
+    // Call OPENSSL_init_crypto to initialize OPENSSL_INIT_LOAD_CONFIG to
+    // avoid the default behavior where errors raised during the parsing of the
+    // OpenSSL configuration file are not propagated and cannot be detected.
+    //
+    // If FIPS is configured the OpenSSL configuration file will have an
+    // .include pointing to the fipsmodule.cnf file generated by the openssl
+    // fipsinstall command. If the path to this file is incorrect no error
+    // will be reported.
+    //
+    // For Node.js this will mean that EntropySource will be called by V8 as
+    // part of its initialization process, and EntropySource will in turn call
+    // CheckEntropy. CheckEntropy will call RAND_status which will now always
+    // return 0, leading to an endless loop and the node process will appear to
+    // hang/freeze.
+    std::string env_openssl_conf;
+    credentials::SafeGetenv("OPENSSL_CONF", &env_openssl_conf);
+
+    bool has_cli_conf = !per_process::cli_options->openssl_config.empty();
+    if (has_cli_conf || !env_openssl_conf.empty()) {
+      OPENSSL_INIT_SETTINGS* settings = OPENSSL_INIT_new();
+      OPENSSL_INIT_set_config_file_flags(settings, CONF_MFLAGS_DEFAULT_SECTION);
+      if (has_cli_conf) {
+        const char* conf = per_process::cli_options->openssl_config.c_str();
+        OPENSSL_INIT_set_config_filename(settings, conf);
+      }
+      OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CONFIG, settings);
+      OPENSSL_INIT_free(settings);
+
+      if (ERR_peek_error() != 0) {
+        result.exit_code = ERR_GET_REASON(ERR_peek_error());
+        result.early_return = true;
+        fprintf(stderr, "OpenSSL configuration error:\n");
+        ERR_print_errors_fp(stderr);
+        return result;
+      }
     }
-  }
 #else
-  if (FIPS_mode()) {
-    OPENSSL_init();
-  }
+    if (FIPS_mode()) {
+      OPENSSL_init();
+    }
 #endif
-  // V8 on Windows doesn't have a good source of entropy. Seed it from
-  // OpenSSL's pool.
-  V8::SetEntropySource(crypto::EntropySource);
+    // V8 on Windows doesn't have a good source of entropy. Seed it from
+    // OpenSSL's pool.
+    V8::SetEntropySource(crypto::EntropySource);
 #endif  // HAVE_OPENSSL
+}
+  if (init_flags & kInitializeV8) {
+    per_process::v8_platform.Initialize(
+        static_cast<int>(per_process::cli_options->v8_thread_pool_size));
```

Comment:
I’d still keep these separate:

```suggestion
  per_process::v8_platform.Initialize(
      static_cast<int>(per_process::cli_options->v8_thread_pool_size));
  if (init_flags & kInitializeV8) {
```

because whether the first call does anything or not is already controlled through `NODE_USE_V8_PLATFORM`